app/src/main/java/com/eveningoutpost/dexdrip/cgm/carelinkfollow/client/CareLinkClient.java

```diff
@@ -0,0 +1,656 @@
+        }
+        return recentData;
+
+    }
+
+    // General data request for API calls
+    protected <T> T getData(HttpUrl url, RequestBody requestBody, Class<T> dataClass) {
+
+        Request.Builder requestBuilder = null;
+        HttpUrl.Builder urlBuilder = null;
+        String authToken = null;
+        String responseString = null;
+        Response response = null;
+        Object data = null;
+
+        this.lastDataSuccess = false;
+        this.lastErrorMessage = "";
+
+        // Get auth token
+        authToken = this.getAuthorizationToken();
+
+        if (authToken != null) {
+
+            // Create request for URL with authToken
+            requestBuilder = new Request.Builder().url(url).addHeader("Authorization", authToken);
+
+            // Add header
+            if (requestBody == null) {
+                this.addHttpHeaders(requestBuilder, RequestType.Json);
+            } else {
+                requestBuilder.post(requestBody);
+                this.addHttpHeaders(requestBuilder, RequestType.HtmlPost);
+            }
+
+            // Send request
+            try {
+                response = this.httpClient.newCall(requestBuilder.build()).execute();
+                this.lastResponseCode = response.code();
+                if (response.isSuccessful()) {
+                    try {
+                        responseString = response.body().string();
+                        data = new GsonBuilder().create().fromJson(responseString, dataClass);
+                        this.lastDataSuccess = true;
+                    } catch (Exception e){
+                        lastErrorMessage = e.getClass().getSimpleName() + ":" + e.getMessage();
+                    }
+                }
+                response.close();
+            } catch (Exception e) {
+                lastErrorMessage = e.getClass().getSimpleName() + ":" + e.getMessage();
+            }
+
+        }
+
+        //Return result
+        if(data != null)
+            return dataClass.cast(data);
+        else
+            return null;
+
+    }
+
+    protected <T> T getData(String host, String path, Map<String, String> queryParams, RequestBody requestBody,
+                            Class<T> dataClass) {
+
+        HttpUrl.Builder urlBuilder = null;
+        HttpUrl url = null;
+
+        // Build url
+        urlBuilder = new HttpUrl.Builder().scheme("https").host(host).addPathSegments(path);
+        if (queryParams != null) {
+            for (Map.Entry<String, String> param : queryParams.entrySet()) {
+                urlBuilder.addQueryParameter(param.getKey(), param.getValue());
+            }
+        }
+
+        url = urlBuilder.build();
+
+        return this.getData(url, requestBody, dataClass);
+
+    }
+
+    // Http header builder for requests
+    protected void addHttpHeaders(Request.Builder requestBuilder, RequestType type) {
+
+        //Add common browser headers
```

Comment:
Are the request headers, especially the user agent really necessary? If yes, xDrip can be blocked easily based on these strings.